funsor/interpreter.py

```diff
@@ -87,39 +90,166 @@ def reinterpret_funsor(x):
+@recursion_reinterpret.register(float)
+@recursion_reinterpret.register(type)
+@recursion_reinterpret.register(types.FunctionType)
+@recursion_reinterpret.register(types.BuiltinFunctionType)
+@recursion_reinterpret.register(torch.Tensor)
+@recursion_reinterpret.register(Domain)
+@recursion_reinterpret.register(Op)
+def recursion_reinterpret_ground(x):
     return x
 
 
-@reinterpret.register(tuple)
-def _reinterpret_tuple(x):
+@recursion_reinterpret.register(tuple)
+def recursion_reinterpret_tuple(x):
     return tuple(map(reinterpret, x))
 
 
-@reinterpret.register(frozenset)
-def _reinterpret_frozenset(x):
+@recursion_reinterpret.register(frozenset)
+def recursion_reinterpret_frozenset(x):
     return frozenset(map(reinterpret, x))
 
 
-@reinterpret.register(dict)
-def _reinterpret_dict(x):
+@recursion_reinterpret.register(dict)
+def recursion_reinterpret_dict(x):
     return {key: reinterpret(value) for key, value in x.items()}
 
 
-@reinterpret.register(OrderedDict)
-def _reinterpret_ordereddict(x):
+@recursion_reinterpret.register(OrderedDict)
+def recursion_reinterpret_ordereddict(x):
     return OrderedDict((key, reinterpret(value)) for key, value in x.items())
 
 
+@singledispatch
+def children(x):
+    raise ValueError(type(x))
+
+
+# has to be registered in terms.py
+def children_funsor(x):
+    return x._ast_values
+
+
+@children.register(tuple)
+@children.register(frozenset)
+def _children_tuple(x):
+    return x
+
+
+@children.register(dict)
+@children.register(OrderedDict)
+def _children_tuple(x):
+    return x.values()
+
+
+@children.register(str)
+@children.register(int)
+@children.register(float)
+@children.register(type)
+@children.register(types.FunctionType)
+@children.register(types.BuiltinFunctionType)
+@children.register(torch.Tensor)
+@children.register(Domain)
+@children.register(Op)
+def _children_ground(x):
+    return ()
+
+
+def is_atom(x):
+    if isinstance(x, (tuple, frozenset)) and not isinstance(x, Domain):
+        return len(x) == 0 or all(is_atom(c) for c in x)
+    return isinstance(x, (
+        int,
+        str,
+        float,
+        type,
+        types.FunctionType,
+        types.BuiltinFunctionType,
+        torch.Tensor,
+        Domain,
+        Op
+    ))
+
+
+def gensym():
+    return "V" + str(uuid.uuid4().hex)
+
+
+def stack_reinterpret(x):
+    r"""
+    Overloaded reinterpretation of a deferred expression.
+    This interpreter uses an explicit stack and no recursion but is much slower.
+
+    This handles a limited class of expressions, raising
+    ``ValueError`` in unhandled cases.
+
+    :param x: An input, typically involving deferred
+        :class:`~funsor.terms.Funsor` s.
+    :type x: A funsor or data structure holding funsors.
+    :return: A reinterpreted version of the input.
+    :raises: ValueError
+    """
+    node_vars = {}
+    env = {}
+    x_name = gensym()
+    stack = [(x_name, x)]
+    parent_to_children = OrderedDict()
+    child_to_parent = OrderedDict()
```

Comment:
Is there a reason you're treating funsors as trees rather than dags? I would have expected dag processing along the lines of 
```py
names = {x_name: x}
parents_to_children = OrderedDict()
children_to_parents = OrderedDict()
while stack:
    ...
    for c in children(h):
        if c in names:
            c_name = names[c]
        else:
            c_name = gensym()
            names[c] = c_name
            stack.append((c_name, c))
        children_to_parents.setdefault(c_name, []).append(h_name)
        parents_to_children[h_name].append(c_name)
```
IIUC these approaches can differ exponentially, e.g.
```py
def example(depth=100):
    x = Variable('x')
    for _ in range(depth):
        x = x * x
    return x 
```

Reply:
My first attempt at this wasn't hashing or refcounting correctly so I started with trees but I've fixed it now